server/src/main/java/org/apache/cassandra/sidecar/job/RepairJob.java

```diff
@@ -0,0 +1,249 @@
+
+        int cmd = storageOperations.repair(keyspace, options);
+        if (cmd <= 0)
+        {
+            // repairAsync can only return 0 for replication factor 1.
+            LOGGER.info("Replication factor is 1. No repair is needed for keyspace '{}'", keyspace);
+        }
+        else
+        {
+            // complete the max wait time promise either when exceeding the wait time, or the result is available
+            Promise<Boolean> maxWaitTimePromise = Promise.promise();
+            vertx.setTimer(repairConfiguration.maxRepairJobRuntimeMillis(), d -> {
+                LOGGER.info("Timer Poll");
+                maxWaitTimePromise.tryComplete(true);
+            });
+
+            // Promise for completion of repair operation
+            Promise<Void> promise = Promise.promise();
+            Future<Void> resultFut = promise.future();
+
+            // main event loop checks periodically (10s) for completion
+            vertx.setPeriodic(repairConfiguration.repairPollIntervalMillis(), id -> queryForCompletedRepair(promise, cmd));
+            resultFut.onComplete(res -> maxWaitTimePromise.tryComplete(false));
+            Future<Boolean> maxWaitTimeFut = maxWaitTimePromise.future();
+
+            Future<Void> compositeFut = Future.any(maxWaitTimeFut, resultFut)
+                                              // If this lambda below is evaluated, either one of the futures have completed;
+                                              // In either case, the future corresponding to the job execution is returned
+                                              .compose(f -> {
+                                                  LOGGER.info("One of the futures ended waitStatus={} resultStatus={}",
+                                                              maxWaitTimeFut.isComplete(), resultFut.isComplete());
+                                                  boolean isTimeout = (maxWaitTimeFut.succeeded()) ? maxWaitTimeFut.result() : false;
+                                                  if (isTimeout)
+                                                  {
+                                                      LOGGER.error("Timer ran out before the repair job completed. Repair took too long");
+                                                      // TODO: Cancel repair? (Nice to have)
+                                                      // We free up the thread (job fails) and stop polling for completion
+                                                      return Future.failedFuture("Repair job taking too long");
+                                                  }
+                                                  // otherwise, the result of the job is available
+                                                  return resultFut;
+                                              });
+            try
+            {
+                compositeFut.toCompletionStage().toCompletableFuture().get(); // wait
```

Comment:
periodic task will keep on running if the job has timedout. Need to be cancelled?